gusto/time_discretisation.py

```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
         self.x1.assign(x_in[1])
         solver.solve()
         x_out.assign(self.x_out)
+
+
+class TR_BDF2(TimeDiscretisation):
+    """
+    Implements the two stage implicit TR-BDF2 timestepping method, with a trapeziodal stage follwed
+    by a second order backwards difference stage.
+
+    The TR_BDF2 timestepping method for operator F is written as
+    y^(n+g) = y^n + dt*g*F[y^n] + dt*g*F[y^(n+g)] (TR stage)
+    y^(n+1) = 1/(g(2-g))*y^(n+g) - (1-g)**2/(g(2-g))*y^(n) + (1-g)/(2-g)*dt*F[y^(n+1)] (BDF2 stage)
+    for parameter g (gamma).
+    """
+    def __init__(self, domain, field_name=None, gamma=None,
+                 solver_parameters=None, options=None):
+        """
+        Args:
+            domain (:class:`Domain`): the model's domain object, containing the
+                mesh and the compatible function spaces.
+            field_name (str, optional): name of the field to be evolved.
+                Defaults to None.
+            gamma (float, optional): the off-centring parameter
+            solver_parameters (dict, optional): dictionary of parameters to
+                pass to the underlying solver. Defaults to None.
+            options (:class:`AdvectionOptions`, optional): an object containing
+                options to either be passed to the spatial discretisation, or
+                to control the "wrapper" methods, such as Embedded DG or a
+                recovery method. Defaults to None.
+
+        Raises:
+            ValueError: if gamma is not provided.
+        """
+        # TODO: would this be better as a non-optional argument? Or should the
+        # check be on the provided value?
+        if gamma is None:
+            raise ValueError("please provide a value for gamma between 0 and 1")
+        if isinstance(options, (EmbeddedDGOptions, RecoveryOptions)):
+            raise NotImplementedError("Only SUPG advection options have been implemented for this time discretisation")
+        if not solver_parameters:
+            # theta method leads to asymmetric matrix, per lhs function below,
```

Comment:
I spotted this below. Presumably you have taken the default solver parameters from the theta method? At the very least we should update the reference to theta method here, but maybe we should think about what our default solver options should be

Reply:
I agree.. I just went with the theta method ones as I hadn't given it much thought. What options are available?